docs/architecture/adr-101-data-companion-pull-api.md

```diff
@@ -0,0 +1,403 @@
+message GetLatestHeightResponse {
+    // The height of the latest committed block. Will be 0 if no data has been
+    // committed yet.
+    uint64 height = 1;
+}
+
+message GetByHeightRequest {
+    // The height of the block to get. Set to 0 to return the latest block.
+    uint64 height = 1;
+}
+
+message GetByHeightResponse {
+    // The ID associated with the relevant block.
+    tendermint.types.BlockID block_id = 1;
+    // Block data for the requested height.
+    tendermint.types.Block block = 2;
+}
+```
+
+#### Block Results Service
+
+The following `BlockResultsService` service is proposed _separately_ to the
+`BlockService`. There are several reasons as to why there are two separate gRPC
+services to meet the companion's needs as opposed to just one:
+
+1. The quantity of data stored by each is application-dependent, and coalescing
+   the two types of data could impose significant overhead in some cases (this
+   is primarily a justification for having separate RPC calls for each type of
+   data).
+2. Operators can enable/disable these services independently of one another.
+3. The existing JSON-RPC API distinguishes between endpoints providing these two
+   types of data (`/block` and `/block_results`), so users are already
+   accustomed to this distinction.
+4. Eventually, when we no longer need to store block results at all, we can
+   simply deprecate the `BlockResultsService` without affecting clients who rely
+   on `BlockService`.
+
+```protobuf
+syntax = "proto3";
+
+package tendermint.services.block_results.v1;
+
+// BlockResultsService provides information about the execution results for
+// specific blocks.
+service BlockResultsService {
+    // GetByHeight attempts to retrieve the execution results associated with a
+    // block of a certain height.
+    rpc GetByHeight(GetByHeightRequest) returns (GetByHeightResponse)
+}
+
+message GetByHeightRequest {
+    // The height of the block whose results are to be retrieved. Set to 0 to
+    // return the latest block's results.
+    uint64 height = 1;
+}
+
+message GetByHeightResponse {
+    // The height associated with the block results.
+    uint64 height = 1;
+
+    // The contents of the FinalizeBlock response, which contain block execution
+    // results.
+    tendermint.abci.ResponseFinalizeBlock finalize_block_response = 2;
+}
+```
+
+#### Pruning Service
+
+This gRPC service is the only novel service proposed in this ADR, and
+effectively gives a single external caller (e.g. a data companion) a say in how
+the node prunes its data.
+
+```protobuf
+syntax = "proto3";
+
+package tendermint.services.pruning.v1;
+
+// PruningService provides privileged access to specialized pruning
+// functionality on the CometBFT node to help control node storage.
+service PruningService {
+    // SetBlockRetainHeightRequest indicates to the node that it can safely
+    // prune all block data up to the specified retain height.
```

Comment:
I assume this is inclusive, right? So if the parameter is `H`, we can prune all heights <= `H`.

Reply:
No, the retain heights, as defined in current versions of Comet and TM, are non inclusive. So you will prune **up** to the retain height but that height stays.